tests/sphere_test.cpp

```diff
@@ -0,0 +1,120 @@
+
+using core::Ray;
+using core::RayHitData;
+using core::Vec3;
+using objects::Sphere;
+
+const float kSkipFreq = 1000;
+const Vec3 kVecZero(0, 0, 0);
+
+float deg2rad(float deg) { return 2 * constants::kPi * deg / 360; }
+
+TEST(SphereCollisionTest, RayHitFromOutsideSphere) {
+
+  Sphere sphere(kVecZero, 1);
+  RayHitData hitData;
+  Ray alongYAxis(Vec3(0, -4, 0), Vec3(0, 1, 0));
+  ASSERT_TRUE(sphere.hitObject(alongYAxis, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(alongYAxis.origin().magnitude() - sphere.getRadius(),
+                  hitData.time);
+
+  Ray alongYAxisNoHit(Vec3(0, 4, 0), Vec3(0, 1, 0));
+  ASSERT_FALSE(sphere.hitObject(alongYAxisNoHit, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+
+  Ray alongXAxis(Vec3(-4, 0, 0), Vec3(1, 0, 0));
+  ASSERT_TRUE(sphere.hitObject(alongXAxis, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(alongXAxis.origin().magnitude() - sphere.getRadius(),
+                  hitData.time);
+
+  Ray alongXAxisNoHit(Vec3(4, 0, 0), Vec3(1, 0, 0));
+  ASSERT_FALSE(sphere.hitObject(alongYAxisNoHit, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+
+  Ray alongZAxis(Vec3(0, 0, -4), Vec3(0, 0, 1));
+  ASSERT_TRUE(sphere.hitObject(alongZAxis, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(alongZAxis.origin().magnitude() - sphere.getRadius(),
+                  hitData.time);
+
+  Ray alongZAxisNoHit(Vec3(0, 0, 4), Vec3(0, 0, 1));
+  ASSERT_FALSE(sphere.hitObject(alongZAxisNoHit, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+
+  Vec3 randomChosenVec3(12.123, 5.623, 13.235);
+  Ray randomChosenOriginRay(randomChosenVec3,
+                            sphere.getOrigin() - randomChosenVec3);
+  ASSERT_TRUE(sphere.hitObject(randomChosenOriginRay, kSkipFreq, &hitData));
+  ASSERT_NEAR(randomChosenOriginRay.origin().magnitude() - sphere.getRadius(),
+              hitData.time, constants::kAccuracy);
+
+  Ray randomChosenOriginRayNoHit(randomChosenVec3,
+                                 randomChosenVec3 - sphere.getOrigin());
+  ASSERT_FALSE(
+      sphere.hitObject(randomChosenOriginRayNoHit, kSkipFreq, &hitData))
+      << "hit at: " << hitData.collisionPoint();
+
+  // This comes from proportions of the pitagorus triangle where sides satisfy
+  // the equation : 3^2 + 4^2 = 5^5
+  Vec3 originFromAngle(0, -3, -4);
+  Ray rayToSphereOriginFromAngle(originFromAngle,
+                                 sphere.getOrigin() - originFromAngle);
+  ASSERT_TRUE(
+      sphere.hitObject(rayToSphereOriginFromAngle, kSkipFreq, &hitData));
+  ASSERT_FLOAT_EQ(5 - sphere.getRadius(), hitData.time);
+
+  Ray alongXOneHit(Vec3(-5, 0, 1), Vec3(1, 0, 0));
+  ASSERT_TRUE(sphere.hitObject(alongXOneHit, kSkipFreq, &hitData));
+  ASSERT_EQ(5, hitData.time);
+  ASSERT_EQ(Vec3(0, 0, 1), hitData.collisionPoint());
+}
+
+TEST(SphereCollisionTest, RayHitInsideSphere) {
+  Sphere sphere(kVecZero, 1);
+  RayHitData hitData;
+
+  Ray alongXAxis(kVecZero, Vec3(1, 0, 0));
```

Comment:
kVecZero -> sphere.origin() ?

It would be better if you made a bit less trivial data, how about picking something that is inside the unit sphere and writing it down here?

You can later check if it is truly inside.

Reply:
changed. I added a mthod in test fixture class that ccheck if given vec is inside of given sphere. 

>  how about picking something that is inside the unit sphere and writing it down here

now I have non trivial vec that is inside sphere and also ray that is really close to the edge of sphere. I think it will be enough for test inside the sphere ;)